agent/sidecar_service.go

```diff
@@ -0,0 +1,146 @@
+package agent
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+func (a *Agent) sidecarServiceID(serviceID string) string {
+	return serviceID + "-sidecar-proxy"
+}
+
+// sidecarServiceFromNodeService returns a *structs.NodeService representing a
+// sidecar service with all defaults populated based on the current agent
+// config.
+//
+// It assumes the ns has been validated already which means the nested
+// SidecarService is also already validated.It also assumes that any check
+// definitions within the sidecar service definition have been validated if
+// necessary. If no sidecar service is defined in ns, then nil is returned with
+// nil error.
+//
+// The second return argument is a list of CheckTypes to register along with the
+// service.
+//
+// The third return argument is the effective Token to use for the sidecar
+// registration. This will be the same as the token parameter passed unless the
+// SidecarService definition contains a distint one.
+func (a *Agent) sidecarServiceFromNodeService(ns *structs.NodeService, token string) (*structs.NodeService, []*structs.CheckType, string, error) {
+	if ns.Connect.SidecarService == nil {
+		return nil, nil, "", nil
+	}
+
+	// Start with normal conversion from service definition
+	sidecar := ns.Connect.SidecarService.NodeService()
+
+	// Override the ID which must always be consistent for a given outer service
+	// ID. We rely on this for lifecycle management of the nested definition.
+	sidecar.ID = a.sidecarServiceID(ns.ID)
+
+	// Set some meta we can use to disambiguate between service instances we added
+	// later and are responsible for deregistering.
+	if sidecar.Meta != nil {
+		// Meta is non-nil validate it before we add the special key so we can
+		// enforce that user cannot add a consul- prefix one.
+		if err := structs.ValidateMetadata(sidecar.Meta, false); err != nil {
+			return nil, nil, "", err
+		}
+	}
+
+	// Flag this as a sidecar - this is not persisted in catalog but only needed
+	// in local agent state to disambiguate lineage when deregistereing the parent
+	// service later.
+	sidecar.LocallyRegisteredAsSidecar = true
+
+	// See if there is a more specific token for the sidecar registration
+	if ns.Connect.SidecarService.Token != "" {
+		token = ns.Connect.SidecarService.Token
+	}
+
+	// Setup some sane connect proxy defaults.
+	if sidecar.Kind == "" {
+		sidecar.Kind = structs.ServiceKindConnectProxy
+	}
+	if sidecar.Service == "" {
+		sidecar.Service = ns.Service + "-sidecar-proxy"
```

Comment:
can't you use `sidecarServiceID()` in here?

Reply:
Heh good question!

This is the _name_ and not the ID. That's important because they can be different (user defined IDs as you discovered) and ACLs are enforced on service name not ID. So the name of our proxy really needs to be based on the _name_ of the target not the _ID_ of the target to allow prefix-based ACL rules to match both reliably. See recent bug: https://github.com/hashicorp/consul/issues/4619